### PR TITLE
Fix #53181. AssemblyLoadContext incorrectly validates name of dynamiс assembly when returned from Load

### LINKED ISSUE #53181: AssemblyLoadContext incorrectly validates name of dynamic assembly when returned from Load
`AssemblyLoadContext` validates that the assembly returned from `Load` has the same name as the name which was asked for. This validation incorrectly handles dynamic assemblies and will always fail for them.

This is the cause because `ValidateAssemblyNameWithSimpleName` simply casts the returned assembly to `RuntimeAssembly`:
https://github.com/dotnet/runtime/blob/01c9cf6102b525f0b787cfa3fcce72cbee2d6716/src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs#L660

But this doesn't work for assemblies created by `AssemblyBuilder`.

The code should call `GetRuntimeAssembly` instead, which handles this case correctly.

Full repro for both affected public APIs (`Load` and `Resolving`) is here:
https://gist.github.com/vitek-karas/3c8f938dc15730ad00a375d8d745f36a

This is basically a derivative issue from the bug fix in https://github.com/dotnet/runtime/pull/49387.

### PR DESCRIPTION
**Pull request by Ap0k**
See #53181

### CODE REVIEW

**Review by vitek-karas — APPROVED** [review]
Looks good - thanks a lot!

File: src/tests/Loader/AssemblyLoadContext30Extensions/AssemblyLoadContext30Extensions.cs [hunk 1]
```diff
@@ -200,6 +202,32 @@ public static void CustomName()
...
+            AssemblyLoadContext alc = new AssemblyLoadContext($"ALC - {isCollectible}", isCollectible);
+            AssemblyBuilder assemblyBuilder;
+            AssemblyName assemblyName = new AssemblyName($"DynamicAssembly_{Guid.NewGuid():N}");
+
+            using (alc.EnterContextualReflection())
+            {
+                assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(assemblyName, AssemblyBuilderAccess.RunAndCollect);
+            }
+
+            AssemblyLoadContext? context = AssemblyLoadContext.GetLoadContext(assemblyBuilder);
```
Comment (vitek-karas) [inline src/tests/Loader/AssemblyLoadContext30Extensions/AssemblyLoadContext30Extensions.cs:220]: Actually - can you also validate that the assembly is in the `ALC.Assemblies` collection?

Comment (Ap0k) [reply src/tests/Loader/AssemblyLoadContext30Extensions/AssemblyLoadContext30Extensions.cs:221]: Added check

**Comment by vitek-karas** [comment]
The test failure in CI is https://github.com/dotnet/runtime/issues/48851.

**Comment by vitek-karas** [comment]
@Ap0k thanks a lot for the fix!

**Pull request merged**